src/websocket.js

```diff
@@ -0,0 +1,47 @@
+import mitt from 'mitt';
+import { encode } from './encoding';
+
+export default (url, options) => {
+  const emitter = mitt();
+
+  emitter.on('data', message => {
+    // need to access the message data
+    const lines = encode(message);
+
+    emitter.emit('update', { lines });
+  });
+
+  emitter.on('start', () => {
+    try {
+      // try to connect to websocket
+      const socket = new WebSocket(url);
+
+      socket.addEventListener('open', event => {
+        if (options.onOpen) {
+          options.onOpen(event, socket);
+        }
+      });
+      socket.addEventListener('close', options.onClose);
+
+      socket.addEventListener('error', error => {
+        emitter.emit('error', error);
+      });
+
+      socket.addEventListener('message', event => {
+        let message = event.data;
+
+        if (options.formatMessage) {
+          message = options.formatMessage(message);
+        }
+
+        emitter.emit('data', message);
+      });
+
+      emitter.on('abort', () => socket.close());
+    } catch (err) {
+      emitter.emit('error', err);
+    }
+  });
+
+  return emitter;
+};
```

Comment:
This is really nice. Thoughts about the following diff + the comment from above:

```diff
diff --git a/src/websocket.js b/src/websocket.js
index d6079a2..c7dcc83 100644
--- a/src/websocket.js
+++ b/src/websocket.js
@@ -13,31 +13,39 @@ export default (url, options) => {
 
   emitter.on('start', () => {
     try {
+      const { onOpen, onClose, formatMessage } = options;
+
       // try to connect to websocket
       const socket = new WebSocket(url);
 
+      // websocket event 1
       socket.addEventListener('open', event => {
-        if (options.onOpen) {
-          options.onOpen(event, socket);
+        if (onOpen) {
+          onOpen(event, socket);
         }
       });
-      socket.addEventListener('close', options.onClose);
-
+      // websocket event 2
+      socket.addEventListener('close', event => {
+        if (onClose) {
+          onClose(event);
+        }
+      });
+      // websocket event 3
       socket.addEventListener('error', error => {
         emitter.emit('error', error);
       });
-
+      // websocket event 4
       socket.addEventListener('message', event => {
         let message = event.data;
 
-        if (options.formatMessage) {
-          message = options.formatMessage(message);
+        if (formatMessage) {
+          message = formatMessage(message);
         }
 
         emitter.emit('data', message);
       });
 
-      emitter.on('abort', () => socket.close());
+      emitter.on('abort', socket.close);
     } catch (err) {
       emitter.emit('error', err);
     }

```